integration/experiment/power_sweep/gen_policy_recommendation.py

```diff
@@ -0,0 +1,287 @@
+        rt_at_tdp = rtmodel.test(pltdp)
+        constrained_values = [
+                (en, pl, rt)
+                for pl, rt, en in zip(plrange, rt_predictions, en_predictions)
+                if rt <= (1 + max_degradation) * rt_at_tdp]
+        besten, bestpl, bestrt = min(constrained_values)
+        return bestpl, bestrt, besten
+
+
+def main(argv):
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--path', help='path containing reports and machine.json', required=True)
+    parser.add_argument('--region_filter', help='comma-separated list of regions to include, default to include all regions', default=None)
+    parser.add_argument('--dump_prefix', help='prefix to dump statistics to, empty to not dump stats')
+    parser.add_argument('--min_pl', help='minimum power limit, default behavior is to use machine.json', default=-1, type=float)
+    parser.add_argument('--max_pl', help='maximum power limit, default behavior is to use machine.json', default=-1, type=float)
+    parser.add_argument('--tdp', help='tdp (sticker power limit), default behavior is to use machine.json', default=-1, type=float)
+    parser.add_argument('--max_degradation', help='maximum allowed runtime degradation, default is 0.1 (i.e., 10%%)', default=0.1, type=float)
+    parser.add_argument('--min_energy', help='ignore max degradation, just give the minimum energy possible', action='store_true')
+
+    args = parser.parse_args(argv[1:])
+
+    simplified_data = extract_from_reports(args.path, args.region_filter)
+    if args.dump_prefix:
+        dump_simplified_data(simplified_data, "%s.dat" % dump_prefix)
+        dump_stats_summary(simplified_data, "%s.stats" % dump_prefix)
+
+    rtmodel = CubicPowerModel()
+    enmodel = CubicPowerModel()
+
+    analyze_extracted_data(simplified_data, rtmodel, enmodel)
+
+    min_pl, tdp, max_pl = None, None, None
+    if min(args.min_pl, args.tdp, args.max_pl) < 0:
+        try:
+            machine_info = json.load(open(os.path.join(args.path, "machine.json")))
+            min_pl = machine_info['signals']['POWER_PACKAGE_MIN']
+            tdp = machine_info['signals']['POWER_PACKAGE_TDP']
+            max_pl = machine_info['signals']['POWER_PACKAGE_TDP']
+        except IOError:
+            sys.stderr.write("Warning: couldn't open machine.json. Falling back to default values.\n")
+            min_pl = 150
+            tdp = 280
+            max_pl = 280
+
+    # user-provided arguments override
+    if args.min_pl > 0:
+        min_pl = args.min_pl
+    if args.tdp > 0:
+        tdp = args.tdp
+    if args.max_pl > 0:
+        max_pl = args.max_pl
+
+    plrange = [min_pl + i for i in range(int(max_pl - min_pl))] + [max_pl]
+    bestpl, bestrt, besten = policy_min_energy(plrange, enmodel, rtmodel, tdp, max_degradation=(None if args.min_energy else args.max_degradation))
+    tdprt, tdpen = rtmodel.test(tdp), enmodel.test(tdp)
+
+    print("AT TDP = %dW, RUNTIME = %f s,  ENERGY = %f J" % (tdp, tdprt, tdpen))
```

Comment:
Please don't use ```print```.  Use only ```sys.stdout.write``` or ```sys.stderr.write```.

Print is not portable between Python 2 and 3.  We're trying to keep ```print``` out of the codebase.

Reply:
mea culpa